app/src/main/java/org/mozilla/fenix/settings/search/EditCustomSearchEngineFragment.kt

```diff
@@ -0,0 +1,180 @@
+        val decodedUrl = Uri.decode(searchEngine.buildSearchUrl("%s"))
+        edit_search_string.setText(decodedUrl)
+
+        custom_search_engines_learn_more.increaseTapArea(DPS_TO_INCREASE)
+        custom_search_engines_learn_more.setOnClickListener {
+            requireContext().let { context ->
+                val intent = SupportUtils.createCustomTabIntent(
+                    context,
+                    SupportUtils.getSumoURLForTopic(
+                        context,
+                        SupportUtils.SumoTopic.CUSTOM_SEARCH_ENGINES
+                    )
+                )
+                startActivity(intent)
+            }
+        }
+    }
+
+    override fun onResume() {
+        super.onResume()
+        (activity as AppCompatActivity).title = getString(R.string.search_engine_edit_custom_search_engine_title)
+        (activity as AppCompatActivity).supportActionBar?.show()
+    }
+
+    override fun onCreateOptionsMenu(menu: Menu, inflater: MenuInflater) {
+        inflater.inflate(R.menu.edit_custom_searchengine_menu, menu)
+    }
+
+    override fun onOptionsItemSelected(item: MenuItem): Boolean {
+        return when (item.itemId) {
+            R.id.save_button -> {
+                saveCustomEngine()
+                true
+            }
+            else -> super.onOptionsItemSelected(item)
+        }
+    }
+
+    private fun saveCustomEngine() {
+        custom_search_engine_name_field.error = ""
+        custom_search_engine_search_string_field.error = ""
+
+        val name = edit_engine_name.text?.toString() ?: ""
+        val searchString = edit_search_string.text?.toString() ?: ""
+
+        var hasError = false
+        if (name.isEmpty()) {
+            custom_search_engine_name_field.error = resources
+                .getString(R.string.search_add_custom_engine_error_empty_name)
+            hasError = true
+        }
+
+        val existingIdentifiers = requireComponents
+            .search
+            .provider
+            .allSearchEngineIdentifiers()
+            .map { it.toLowerCase(Locale.ROOT) }
+
+        val nameHasChanged = name != engineIdentifier
+
+        if (existingIdentifiers.contains(name.toLowerCase(Locale.ROOT)) && nameHasChanged) {
+            custom_search_engine_name_field.error = resources
+                .getString(R.string.search_add_custom_engine_error_existing_name, name)
+            hasError = true
+        }
+
+        if (searchString.isEmpty()) {
+            custom_search_engine_search_string_field
+                .error = resources.getString(R.string.search_add_custom_engine_error_empty_search_string)
+            hasError = true
+        }
+
+        if (!searchString.contains("%s")) {
+            custom_search_engine_name_field
+                .error = resources.getString(R.string.search_add_custom_engine_error_missing_template)
+            hasError = true
+        }
+
+        if (hasError) { return }
+
+        viewLifecycleOwner.lifecycleScope.launch {
```

Comment:
I'd prefer the entire function be a suspend function, and we call launch in `onOptionsItemSelected`

Reply:
:+1: